velib_spot_predictor/data/publish.py

```diff
@@ -0,0 +1,550 @@
+        n_columns = len(column.iloc[0])
+        for i in range(n_columns):
+            name = list(column.iloc[0][i].keys())[0]
+            column_name = f"{column.name}_{name}"
+            flattened_column[column_name] = column.str[i].str[name]
+
+        return flattened_column
+
+    @staticmethod
+    def clean_data(data: pd.DataFrame) -> pd.DataFrame:
+        """Clean the data.
+
+        Parameters
+        ----------
+        data : pd.DataFrame
+            Data to clean
+
+        Returns
+        -------
+        pd.DataFrame
+            Cleaned data
+        """
+        column_to_flatten = ["num_bikes_available_types"]
+        for column in column_to_flatten:
+            flattened_column = JsonToSQLBase._flatten_column(data[column])
+            data = pd.concat([data, flattened_column], axis=1)
+        data = data.drop(columns=column_to_flatten)
+        return data
+
+    @staticmethod
+    def extract_datetime_from_filename(filename: str) -> pd.Timestamp:
+        """Extract the datetime from the filename.
+
+        Parameters
+        ----------
+        filename : str
+            Filename to extract the datetime from
+
+
+        Returns
+        -------
+        pd.Timestamp
+            Extracted datetime
+        """
+        match = re.search(r"(\d{8}-\d{6})", filename)
+        if match:
+            datetime_str = match.group(0)
+            return pd.Timestamp(datetime_str).round("min")
+        else:
+            raise ValueError("Invalid filename format")
+
+
+## Extractors
+
+
+class DataFrameExtractor(BaseModel, IExtractor):
+    """Simple extractor to extract the data from an input dataframe."""
+
+    data: list
+    timestamp: datetime = Field(
+        default_factory=lambda: pd.Timestamp.now().floor("min")
+    )
+
+    @field_validator("timestamp", mode="after")
+    @classmethod
+    def check_rounding(cls, timestamp: datetime):
+        """Check that the timestamp is rounded to the minute."""
+        return pd.Timestamp(timestamp).floor("min")
+
+    model_config = {
+        "arbitrary_types_allowed": True,
+    }
+
+    def extract(self) -> pd.DataFrame:
+        """Extract the data.
+
+        Returns
+        -------
+        pd.DataFrame
+            Extracted data
+        """
+        return JsonToSQLBase.clean_data(
+            pd.read_json(StringIO(json.dumps(self.data)))
+        ).assign(datetime=self.timestamp)
+
+
+class FolderExtractor(BaseModel, IExtractor):
+    """Extract the data from the raw data folder.
+
+    Parameters
+    ----------
+    folder_raw_data : DirectoryPath
+        Folder containing the raw data
+    pattern_raw_data : str, optional
+        Pattern to match the raw data files, by default "*.json"
+    pbar : bool, optional
+        Whether to display a progress bar, by default False
+    """
+
+    folder_raw_data: DirectoryPath
+    pattern_raw_data: str = "*.json"
+    pbar: bool = False
+
+    def _find_filepath(self, date: datetime) -> str:
```

Comment:
La méthode n'est jamais utilisée, elle est aussi la seule à utiliser la fonction 'get_one_filepath' dans le fichier utils et semble donc inutile. tu as peut être pour objectif de t'en servir à un moment ?

Reply:
Actually, this function is used by `FolderToSQLETL` in `velib_spot_predictor.data.cli` in `load_to_sql` command which is accessible with the command `load_data_from_local_to_sql` directly in the environment. The inheritance is as follows:
- `FolderToSQLETL` inherits from
- `BaseFolderETL` which implements
- `FolderExtractor` as its extractor

Reply:
I wanted to mention that only the _find_filepath method is not used, but the FolderExtractor class is still useful. What I see in data.cli.load_to_sql is that FolderToSQLETL calls .run that calls .extract of FolderExtractor which never uses _find_filepath. making this method and the utils.get_one_filepath unused

Reply:
Removed